### DISCUSSION TITLE
NULL pixel - Is there a way to know how many pixels are assigned NULL after applying "camtrim" function? And is there a way to output the NULL / Non-NULL image coordinate to a list?

### DISCUSSION THREAD

**Discussion by skyrockets-21** [DOI-USGS/ISIS3#5605] [General]
As titled. Thank you!

**Comment by Kelvinrr** [comment]
To get the counts you can use [stats](https://isis.astrogeology.usgs.gov/8.1.0/Application/presentation/Tabbed/stats/stats.html).

To get exact locations might be trickier, as far as I know there isn't an ISIS app that allows for this. You could use something like [plio](https://github.com/DOI-USGS/plio) to get that information, but that is a bit more advanced python scripting. If you're interested in going that route I can give some code examples. Otherwise, I'll see if others at Astro know of a way to accomplish this via ISIS apps.

**Reply by Kelvinrr** [reply]
It took a minute so here is an example: https://gist.github.com/Kelvinrr/990329c9b14e50b3040c4a478abd6828

you would need a conda environment with plio and gdal (`condo create -n plio plio gdal`), if that seems too much consider opening an issue to add this feature to ISIS's tools.

**Comment by thareUSGS** [comment]
For the Non-NULL (and maybe NULL pixels), there is a 3rd party Python script, by Alessandro Frigeri, called [pix2vect](https://github.com/afrigeri/pix2vec) which simply wraps the ISIS routine _campt_. The routine was built to enable the conversion from raster pixels to vector polygons, which can help visualize oblique images or fly-by images (where the pixels are not always the same size in the image).

So, while the routine pix2vect saves out all sides of every pixel, the code could be easily commented to just plot the information for the center of each pixel. I think you could simply comment out lines 130-134 (and maintain [line 135](https://github.com/afrigeri/pix2vec/blob/0718d375dbd53a459a3fef6441eabceb9ddb279d/pix2vec/pix2vec.py#L135)), which should also speed up the run.

So as a python-wrapped routine hitting _campt_, it is very slow. As a much faster version, Alessandro just submitted a [C++ update to pixel2map](https://github.com/DOI-USGS/ISIS3/pull/5606) to the ISIS Team. While it is still under review, if this version is approved, it might be released with a future version of ISIS. This C++ version is much much faster. To use this update of pix2map now, one would need to compile ISIS themselves (with his updates).

**Comment by rbeyer** [comment]
Seems like you'd be better off using something like _rasterio_ to just query the pixel locations of NULL data pixels and extract their world coordinates (but that requires being able to write a little Python).

**Reply by thareUSGS** [reply]
_rasterio_ might not pick up all the special pixel values (outside of NULL)? GDAL likes to build a mask now for ISIS3 cubes because of the other special pixel values. Not sure what _rasterio_ will do in that case. 

How about just trying _isis2ascii_ to extract the coordinates to a list?
https://isis.astrogeology.usgs.gov/8.2.0/Application/presentation/Tabbed/isis2ascii/isis2ascii.html

**Reply by rbeyer** [reply]
Yeah, it might not, but rasterio is built on GDAL, so it might.  The _kalasiris_ [specialpixels.py](https://github.com/rbeyer/kalasiris/blob/main/kalasiris/specialpixels.py) module can help manage those if you need to stay in Python, but if you aren't already using _rasterio_ and _kalasiris_, then Kelvin's solution with _plio_ seems good.

**Reply by Kelvinrr** [reply]
I only avoided kalasiris because it's another dependency and idk if @skyrockets-21 would be comfortable with more complexity. So I just hard coded ISISNULL in the example.
